### ISSUE TITLE
`TokenDocument#_onUpdateBaseActor` expansion

### ISSUE THREAD

**Issue by stwlam** [foundryvtt/foundryvtt#8812]
### User Experience

Currently, `TokenDocument#_onUpdateBaseActor` potentially triggers the re-renders/re-draws of the following:
- Combat tracker
- Token attribute bars
- Effect icons and overlay

The pf2e system has expanded this to redraw virtually any other aspect of the token, such as dimensions, scale, `texture.src`, etc. Vision and detection mode changes are also fair game. For linked tokens, `TokenDocumentPF2e#_onUpdateBaseActor` may also call a token data reset, as `ActorPF2e` has data that is pulled by `TokenDocumentPF2e` during its own data preparation.

Finally, a companion `TokenDocumentPF2e#onActorEmbeddedItemChange` method was added, to be called when a (linked) token's actor receives changes to its embedded items. It was originally part of `TokenDocumentPF2e#_onUpdateBaseActor`, but there was a hazard of triggering partial asynchronous `Token` redraws in parallel with the parent `#_onUpdateBaseActor` method that could lead to exception throws at the PIXI level.

With the above, we've created a moderate maintenance burden for ourselves, as canvas rendering is a frequent target for changes in core Foundry updates. It would be great were the following to be considered:
- Call `TokenDocument#_onUpdateBaseActor` from `Actor#_onEmbeddedDocumentChange` or otherwise have another solution to inform `TokenDocument`s of changes to their actors' embedded items.
- Perform a data reset of a `TokenDocument` following the same from its `Actor`.
- Either always redraw the full token when its actor is updated (including item changes) or at least be prepared to do: perhaps there could be some kind of API for cuing what may need to be redrawn.

**Comment by stwlam** [comment]
As a small, contained demonstration in dnd5e, I added the following to `TokenDocument5e`:
```js
  /** @inheritdoc */
  prepareBaseData() {
    super.prepareBaseData();

    if (!this.actorLink) return;

    const dimension = {
      tiny: 0.5,
      sm: 1,
      med: 1,
      lg: 2,
      huge: 3,
      grg: 4,
    }[this.actor?.system.traits.size ?? "med"];

    this.width = this.height = dimension;
  }
```
My dwarf has an active effect that changes its size category to huge:
![image](https://user-images.githubusercontent.com/106829671/214894200-1ad56687-cdb3-4602-9476-8f3ad05814a1.png)
![image](https://user-images.githubusercontent.com/106829671/214894458-ef40e1e2-f786-444a-9333-57ac9da5a52b.png)

Before enabling the AE:
![image](https://user-images.githubusercontent.com/106829671/214894629-ccae6768-9e4e-4222-8848-8f02ba926279.png)

After enabling it and moving the token (since there's no `_onUpdateBaseActor` override to trigger a redraw):
![image](https://user-images.githubusercontent.com/106829671/214894876-9ee92122-931e-4f52-b05a-267384ffc046.png)

**Comment by aaclayton** [comment]
Related to #6892

**Comment by Fyorl** [comment]
Thank you for the detailed breakdown. I've been looking into this with respect to the changes we've made to synthetic actors in v11.
> Call `TokenDocument#_onUpdateBaseActor` from `Actor#_onEmbeddedDocumentChange` or otherwise have another solution to inform `TokenDocument`s of changes to their actors' embedded items.

This requirement makes sense, and should be possible to support.

> Perform a data reset of a `TokenDocument` following the same from its `Actor`.

The motivations for this are also pretty clear with the example you gave, however, not every system is going to have such a data dependency between its Tokens and Actors. For those that don't have such a dependency, resetting the Token every time its Actor changes is unnecessary overhead. Is this something that could be handled on a system-by-system basis, or is there a gap in the API surface that makes this difficult to do?

> Either always redraw the full token when its actor is updated (including item changes) or at least be prepared to do: perhaps there could be some kind of API for cuing what may need to be redrawn.

I believe this should be possible with the new 'render flags' API. For example, you could call `this.object.renderFlags.set({refreshSize: true})` in `TokenDocument` if you know the Token's size needs to be redrawn.

**Comment by stwlam** [comment]
> The motivations for this are also pretty clear with the example you gave, however, not every system is going to have such a data dependency between its Tokens and Actors. For those that don't have such a dependency, resetting the Token every time its Actor changes is unnecessary overhead. Is this something that could be handled on a system-by-system basis, or is there a gap in the API surface that makes this difficult to do?

I think we'll be fine still maintaining this small aspect.

**Comment by Fyorl** [comment]
> I think we'll be fine still maintaining this small aspect.

Alright, in that case I will tentatively mark this as completed given the changes we've implemented, but feel free to re-open it or raise a new issue if it's not working as expected.
